msgraphcore/graph_session.py

```diff
@@ -1,21 +1,28 @@
 """
-Creates a session object
+Graph Session
 """
 from requests import Session, Request, Response
 
 from msgraphcore.constants import BASE_URL, SDK_VERSION
 from msgraphcore.middleware._middleware import MiddlewarePipeline, BaseMiddleware
+from msgraphcore.middleware.options.auth_middleware_options import AuthMiddlewareOptions
+from msgraphcore.middleware._base_auth import AuthProviderBase
+from msgraphcore.middleware.authorization_handler import AuthorizationHandler
 
 
 class GraphSession(Session):
     """
     Extends session object with graph functionality
     """
-    def __init__(self, **kwargs):
+    def __init__(self, scopes: [str], auth_provider: AuthProviderBase, middleware: list = []):
         super().__init__()
         self.headers.update({'sdkVersion': 'graph-python-' + SDK_VERSION})
         self._base_url = BASE_URL
-        middleware = kwargs.get('middleware')
+
+        options = AuthMiddlewareOptions(scopes)
+        auth_handler = AuthorizationHandler(auth_provider, auth_provider_options=options)
+
+        middleware.insert(0, auth_handler)
```

Comment:
@MIchaelMainer we'll have this concern if we allow users to pass their own authorization handler but that -- custom user authorization handler -- is not a requirement we have.  The current set up ensures that the authorization handler is always the first middleware.

Reply:
This is Python. My understanding is that the word 'allow' is funny word to use. 

They can pass their own AuthorizationHandler with a different authProvider via the middleware[] since AuthorizationHandler is middleware.

The current setup ensures that the authorization handler is always the first middleware, but it doesn't ensure that it is the only AuthorizationHandler. If we don't want customers to create a custom AuthorizationHandler, we should at least use the convention `class _AuthorizationHandler`. If we do that convention, at least we letting customers know that we don't intend for that class to be publicly used.

https://www.python.org/dev/peps/pep-0008/

*_single_leading_underscore*: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.

Reply:
Having the class with a leading underscore means that the class is **private** within a **module**. We are using this class in the **graph_session** module.  If we mark it as private then we shouldn't use it in `graph_session.py`. As it is right now, we do not expose `AuthorizationHandler` to the user when they do `from msgraphcore import AuthProvider, GraphSession`. 

A user will have to go out of their way to get AuthorizationHandler, which we can't control because of the nature of Python.

Reply:
Now I understand Python module.